ScoutingManager.cpp

```diff
@@ -0,0 +1,246 @@
+#pragma once
+
+#include "Source\ScoutingManager.h"
+
+ScoutingManager::ScoutingManager()
+	: last_overlord_scout_sent_(0),
+	last_zergling_scout_sent_(0),
+
+	initial_scouts_(true),
+	let_overlords_scout_(true),
+	exists_overlord_scout_(false),
+	exists_zergling_scout_(false),
+	exists_expo_zergling_scout_(false),
+	found_enemy_base_(false),
+
+	overlord_scout_(nullptr),
+	zergling_scout_(nullptr),
+	expo_zergling_scout_(nullptr),
+	last_overlord_scout_(nullptr),
+	last_zergling_scout_(nullptr),
+	last_expo_scout_(nullptr),
+
+	scout_start_positions_(NULL),
+	scout_expo_positions_(NULL)
+{
+}
+
+Position ScoutingManager::getScoutTargets(const Unit &unit, Map_Inventory &inv, Unit_Inventory &ei) {
+// Scouting priorities
+	Position scout_spot;
+	Position e_base_scout = Positions::Origin;
+
+	// Get the mean enemy base location if we have found the enemy base
+	if (inv.getMeanEnemyBuildingLocation(ei) != Positions::Origin) {
+		Stored_Unit* nearest_building = CUNYAIModule::getClosestGroundStored(ei, inv.getMeanEnemyBuildingLocation(ei), inv);
+		if (nearest_building != nullptr) {
+			e_base_scout = nearest_building->pos_;
+			found_enemy_base_ = true;
+		}
+	}
+	
+	// If we haven't found any enemy buildings yet
+	if (!found_enemy_base_) {
+
+		// Build our scout start positions when empty, should only build once
+		if (scout_start_positions_.empty()) {
+			for (auto pos : inv.start_positions_) 
+				scout_start_positions_.push_back(pos);
+		}
+
+		scout_spot = scout_start_positions_[0]; 
+		scout_start_positions_.erase(scout_start_positions_.begin());  // Scouts go to unique starting locations
+		return scout_spot;
+	}
+
+	// Found an enemy building
+	if (found_enemy_base_) {
+		
+		// Suicide zergling or an overlord scout
+		if (zergling_scout_ == unit || overlord_scout_ == unit) { 
+			scout_spot = e_base_scout;   // Path into their base
+			return scout_spot;
+		}
+
+		// Check for expos/hidden stuff scout
+		if (expo_zergling_scout_ == unit) {
+			
+			// Build our scout expo positions when empty, will rebuild when empties
+			if (scout_expo_positions_.empty()) {
+
+				int total_distance = 0;
+
+				// Doesn't feel optimal, likely better off using probabilities but was stuck on the implementation
```

Comment:
Not a bad rewrite - it's practical, it will work.  

I spent some time with to make sure I got one set up, try the following algo:
d=distance to expo D=total distance to expo, N=number of unexplored expos.
prob of any one expo being selected = (1-d/D)/(N-1)

Here's a quick test in excel:

  | Distances | (1-d/D)/(N-1)
-- | -- | --
1 | 17.33938 | 0.054446135
2 | 80.20737 | 0.05042367
3 | 50.99409 | 0.052292815
4 | 72.55308 | 0.050913412
5 | 96.59525 | 0.049375129
6 | 63.50811 | 0.051492134
7 | 20.15787 | 0.0542658
8 | 16.38554 | 0.054507164
9 | 18.97983 | 0.054341175
10 | 81.0057 | 0.05037259
11 | 48.13864 | 0.052475515
12 | 52.31649 | 0.052208204
13 | 0.856764 | 0.055500737
14 | 26.92358 | 0.053832912
15 | 61.41087 | 0.051626321
16 | 0.023589 | 0.055554046
17 | 91.61596 | 0.049693717
18 | 17.04425 | 0.054465018
19 | 52.23368 | 0.052213503
Total | 868.29 | 1

This also works, however it does not discount faraway expos as much as we would like. Still, it's a pretty reasonable starting place.